### PR TITLE
fix(tianm): fix problems related to master slave synchronization (#819)

### LINKED ISSUE #818: bug: Master slave synchronization - There will be too many tuples  problem
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```
[2022-10-25 17:43:58.997778] [7781] [ERROR] [tianmu_handler.cpp:1088] MSG:  rnd_pos::position: 808464432
[2022-10-25 17:43:58.997851] [7781] [ERROR] [tianmu_handler.cpp:1089] MSG:  rnd_pos::buf: ÿ3
[2022-10-25 17:43:58.997872] [7781] [ERROR] [tianmu_handler.cpp:1090] MSG:  rnd_pos::pos: a2720000
[2022-10-25 17:43:58.997889] [7781] [ERROR] [tianmu_handler.cpp:1091] MSG:  rnd_pos::ref_length: 8
[2022-10-25 17:43:59.002205] [7783] [ERROR] [tianmu_handler.cpp:1088] MSG:  rnd_pos::position: 201492416
[2022-10-25 17:43:59.002257] [7783] [ERROR] [tianmu_handler.cpp:1089] MSG:  rnd_pos::buf: ÿ3
[2022-10-25 17:43:59.002274] [7783] [ERROR] [tianmu_handler.cpp:1090] MSG:  rnd_pos::pos: 
[2022-10-25 17:43:59.002287] [7783] [ERROR] [tianmu_handler.cpp:1091] MSG:  rnd_pos::ref_length: 8
[2022-10-25 17:43:59.046846] [7781] [WARN] [exception.cpp:42] MSG: Exception: Too many tuples.    (48).

STACK TRACE BEGIN
        /stonedb57_2/install/bin/mysqld(Tianmu::common::OutOfMemoryException::OutOfMemoryException(std::string const&)+0x23) [0x2c6d297]
        /stonedb57_2/install/bin/mysqld(Tianmu::core::Filter::Filter(long, unsigned int, bool, bool)+0x1ae) [0x2c673ba]
        /stonedb57_2/install/bin/mysqld(std::_MakeUniq<Tianmu::core::Filter>::__single_object std::make_unique<Tianmu::core::Filter, unsigned long, unsigned char>(unsigned long&&, unsigned char&&)+0x64) [0x3007860]
        /stonedb57_2/install/bin/mysqld(Tianmu::dbhandler::TianmuHandler::rnd_pos(unsigned char*, unsigned char*)+0x17b) [0x2fffe81]
        /stonedb57_2/install/bin/mysqld(handler::ha_rnd_pos(unsigned char*, unsigned char*)+0x123) [0x1d31727]
        /stonedb57_2/install/bin/mysqld(handler::rnd_pos_by_record(unsigned char*)+0xac) [0x1d4014e]
        /stonedb57_2/install/bin/mysqld(Rows_log_event::do_index_scan_and_update(Relay_log_info const*)+0x37d) [0x2610367]
        /stonedb57_2/install/bin/mysqld(Rows_log_event::do_apply_event(Relay_log_info const*)+0x10de) [0x26127c8]
        /stonedb57_2/install/bin/mysqld(Log_event::do_apply_event_worker(Slave_worker*)+0x7e) [0x2620eb0]
        /stonedb57_2/install/bin/mysqld(Slave_worker::slave_worker_exec_event(Log_event*)+0x49c) [0x2699506]
        /stonedb57_2/install/bin/mysqld(slave_worker_exec_job_group(Slave_worker*, Relay_log_info*)+0x331) [0x269b855]
        /stonedb57_2/install/bin/mysqld(handle_slave_worker+0x42b) [0x267370e]
        /stonedb57_2/install/bin/mysqld(pfs_spawn_thread+0x173) [0x2b9065a]
        /lib64/libpthread.so.0(+0x81a3) [0x7f516f5bd1a3]
        /lib64/libc.so.6(clone+0x6d) [0x7f516ca57fad]
STACK TRACE END

[2022-10-25 17:43:59.047261] [7781] [ERROR] [tianmu_handler.cpp:1112] MSG: An exception is caught: Too many tuples.    (48)
```

### Expected behavior

_No response_

### How To Reproduce

_No response_

### Environment

_No response_

### Are you interested in submitting a PR to solve the problem?

- [X] Yes, I will!

### LINKED ISSUE #819: bug: Master slave synchronization - Primary key conflict problem
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

There will be primary key conflict when modifying data in the case of master slave synchronization

### Expected behavior

_No response_

### How To Reproduce

_No response_

### Environment

_No response_

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by konghaiya**
>fix 1:#818 Master slave synchronization - There will be too many tuples problem

 Cause:
Tianmu::dbhandler::TianmuHandler::current_position This variable is not initialized. In some cases,
 a large value may cause too many tuples problem.
Solution: Initialize the variable
>fix 2:#819 Master slave synchronization - Primary key conflict problem 

Solution: Modify the modification logic of master slave synchronization, so that the delete and update operations do not follow the primary key logic

>Supplement
1. sql/sql_insert.cc 
       Fix the problem that the insert statement will not generate binlog in the delayed insert mode
2. storage/tianmu/handler/tianmu_handler.cpp 
       Fix the binlog error of the line format generated by the tianmu engine

## Summary about this PR

Issue Number: close #818
Issue Number: close #819

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Comment by konghaiya** [comment]
This problem belongs to master-slave synchronization, and mtr cannot cover it. Therefore, no mtr test case is submitted this time.

Verified in offline server

File: sql/log_event.cc [hunk 1]
```diff
@@ -9682,6 +9682,20 @@ search_key_in_table(TABLE *table, MY_BITMAP *bi_cols, uint key_type)
   uint res= MAX_KEY;
   uint key;
 
+  /*
```
Comment (RingsC) [inline sql/log_event.cc:9685]: Tests for primary-secondary scenario should be added. But, we dont see any cases.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Comment by RingsC** [comment]
> This pull request's title should follow requirements next. @konghaiya please check it 👇.
> 
> Valid format:
> 
> ```
> fix(vc): fix sth..... (#3306)
>   ^         ^---------^  ^----^
>   |         |            |
>   |         +            +-> you issue id.
>   |         |
>   |         +-> Summary in present tense.
>   |
>   +-------> Type: feat, fix, docs, workflow, style, refactor, test, website, chore
> ```
> 
> Valid types:
> 
> * `feat`: new feature for stonedb
> * `fix`: bug fix for stonedb
> * `docs`: changes to the documentation
> * `workflow`: ci/cd in .github
> * `refactor`: refactoring production code, eg. renaming a variable
> * `style`: formatting, missing semi colons, etc; no production code change
> * `test`: adding missing tests, refactoring tests; no production code change
> * `website`
> * `chore`: updating grunt tasks etc; no production code change

Do you see this tip, Title sytle should be followed, otherwsie, PR will not be reviewed and merged.

File: sql/log_event.cc (hunk 1 above)
Comment (konghaiya) [reply sql/log_event.cc:9685]: This logic belongs to master slave synchronization logic, and MTR cannot test it.
It has been verified in our test server.

File: sql/log_event.cc [hunk 2]
```diff
@@ -9682,6 +9682,20 @@ search_key_in_table(TABLE *table, MY_BITMAP *bi_cols, uint key_type)
...
+      The primary key of the tianmu engine does not support delete and update statements.
+      The following codes can be deleted after subsequent support
+  */
+  bool tianmu_engine = table && table->s && 
+                      (table->s->db_type() ? table->s->db_type()->db_type == DB_TYPE_TIANMU: false);
+  enum_sql_command sqlCommand = SQLCOM_END;
+  if(table->in_use && table->in_use->lex) sqlCommand = table->in_use->lex->sql_command;
+  bool tianmuDeleteOrUpdate = (tianmu_engine && (sqlCommand == SQLCOM_DELETE ||
```
Comment (RingsC) [inline sql/log_event.cc]: tianmuDeleteOrUpdate ？ if we dont use in the following code, it's not necssary to introduce a new variable.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/log_event.cc (hunk 2 above)
Comment (RingsC) [reply sql/log_event.cc]: The test cases should include innodb table and tianmu table both.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/pack_int.cpp [hunk 3]
```diff
@@ -583,19 +597,22 @@ void PackInt::LoadDataFromFile(system::Stream *fcurfile) {
       ASSERT(!IsModeNoCompression());
 
       uint *cur_buf = (uint *)uptr.get();
-      if (data.ptr == nullptr && data.vt > 0) data.ptr = alloc(data.vt * dpn->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
+      if (data.ptr == nullptr && data.vt > 0)
+        data.ptr = alloc(data.vt * dpn->numOfRecords, mm::BLOCK_TYPE::BLOCK_UNCOMPRESSED);
```
Comment (RingsC) [inline storage/tianmu/core/pack_int.cpp:601]: do memory clear after a new memory allocated.

File: sql/log_event.cc (hunk 1 above)
Comment (konghaiya) [reply sql/log_event.cc:9685]: To add test cases

**Comment by konghaiya** [comment]
Since there is too much difference with the main branch, it is necessary to pull the branch again and re PR, and this PR should be closed first

**Pull request closed without merging**